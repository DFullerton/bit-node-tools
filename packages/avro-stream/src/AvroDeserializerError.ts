import { Message } from 'kafka-node';

export class AvroDeserializerError extends Error {
  constructor(message: string, public chunk: Message, public encoding: string, public originalError?: Error) {
    super(message);
  }
}
