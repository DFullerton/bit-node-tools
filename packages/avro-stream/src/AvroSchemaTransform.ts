import { ForSchemaOptions, Schema, Type } from 'avsc';
import { Transform } from 'stream';
import { SchemaRegistryResolver } from './SchemaRegistryResolver';
import { SchemaResolver } from './types';

export abstract class AvroSchemaTransform extends Transform {
  protected resolver: SchemaResolver;

  constructor(resolver: SchemaResolver | string, private schemaOptions?: Partial<ForSchemaOptions>) {
    super({ objectMode: true });
    this.resolver = typeof resolver === 'string' ? new SchemaRegistryResolver(resolver) : resolver;
  }

  typeForSchema(schema: Schema) {
    return Type.forSchema(schema, { registry: {}, ...this.schemaOptions });
  }
}
