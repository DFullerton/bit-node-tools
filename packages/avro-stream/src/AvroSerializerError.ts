import { AvroProduceRequest } from './types';

export class AvroSerializerError extends Error {
  constructor(
    message: string,
    public chunk: AvroProduceRequest,
    public encoding: string,
    public originalError?: Error,
  ) {
    super(message);
  }
}
