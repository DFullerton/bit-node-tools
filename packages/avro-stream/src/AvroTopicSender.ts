import { Readable, ReadableOptions } from 'stream';
import { AvroProduceRequest, AvroTopicSenderOptions } from './types';

export class AvroTopicSender<TMessage = any> extends Readable {
  readonly options: AvroTopicSenderOptions;

  constructor(options: AvroTopicSenderOptions & ReadableOptions) {
    const { schema, topic, partition, key, ...readableOptions } = options;
    super({ ...readableOptions, objectMode: true });
    this.options = { schema, topic, partition, key };
  }

  send(...messages: TMessage[]) {
    const produceRequest: AvroProduceRequest<TMessage> = { ...this.options, messages };
    this.push(produceRequest);
    return this;
  }

  close() {
    this.push(null);
    return this;
  }
}
