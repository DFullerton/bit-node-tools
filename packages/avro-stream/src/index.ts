export { AvroDeserializer } from './AvroDeserializer';
export { AvroSerializer } from './AvroSerializer';
export { AvroSchemaTransform } from './AvroSchemaTransform';
export { SchemaRegistryResolver } from './SchemaRegistryResolver';
export { SchemaResolver, AvroMessage, AvroProduceRequest } from './types';
export { constructMessage, deconstructMessage } from './message';
