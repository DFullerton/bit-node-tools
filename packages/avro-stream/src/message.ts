export interface AvroMessage {
  schemaId: number;
  buffer: Buffer;
}

export const deconstructMessage = (buffer: Buffer): AvroMessage => {
  return { schemaId: buffer.readInt32BE(1), buffer: buffer.slice(5) };
};

export const constructMessage = ({ schemaId, buffer }: AvroMessage): Buffer => {
  const prefix = Buffer.alloc(5);
  prefix.writeUInt8(0, 0);
  prefix.writeUInt32BE(schemaId, 1);

  return Buffer.concat([prefix, buffer]);
};
