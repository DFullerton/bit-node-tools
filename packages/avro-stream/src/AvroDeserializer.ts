import { Type } from 'avsc';
import { Message } from 'kafka-node';
import { Transform, TransformCallback } from 'stream';
import { deconstructMessage } from './message';
import { SchemaRegistryResolver } from './SchemaRegistryResolver';
import { SchemaResolver } from './types';

export class AvroDeserializer extends Transform {
  private resolver: SchemaResolver;

  constructor(resolver: SchemaResolver | string) {
    super({ objectMode: true });
    this.resolver = typeof resolver === 'string' ? new SchemaRegistryResolver(resolver) : resolver;
  }

  async _transform(message: Message, encoding: string, callback: TransformCallback) {
    try {
      if (typeof message.value === 'string') {
        throw new Error(`ConsumerGroupStream for topic "${message.topic}" must set the encoding to "buffer"`);
      }

      const { schemaId, buffer } = deconstructMessage(message.value);
      const schema = await this.resolver.fromId(schemaId);
      callback(undefined, Type.forSchema(schema).fromBuffer(buffer));
    } catch (error) {
      callback(error);
    }
  }
}
