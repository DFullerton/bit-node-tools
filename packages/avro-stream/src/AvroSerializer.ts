import { TransformCallback } from 'stream';
import { AvroSchemaTransform } from './AvroSchemaTransform';
import { constructMessage } from './message';
import { AvroProduceRequest } from './types';

export class AvroSerializer extends AvroSchemaTransform {
  async _transform(request: AvroProduceRequest, encoding: string, callback: TransformCallback) {
    try {
      const type = this.typeForSchema(request.schema);
      const schemaId = await this.resolver.toId(request.topic, request.schema);

      callback(undefined, {
        ...request,
        messages: request.messages.map(message => constructMessage({ schemaId, buffer: type.toBuffer(message) })),
      });
    } catch (error) {
      callback(error);
    }
  }
}
