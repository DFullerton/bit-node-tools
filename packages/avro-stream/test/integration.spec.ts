import { idToSchema } from '@ovotech/schema-registry-api';
import { Type } from 'avsc';
import { readdirSync, readFileSync } from 'fs';
import { ConsumerGroupStream, KafkaClient, Producer, ProducerStream } from 'kafka-node';
import { join } from 'path';
import { Readable } from 'stream';
import { ReadableMock, WritableMock } from 'stream-mock';
import * as uuid from 'uuid';
import { AvroDeserializer, deconstructMessage } from '../src';
import { AvroSerializer } from '../src';
import { DateType } from './DateType';
import { TimestampType } from './TimestampType';

const createTopics = async (topics: string[]) => {
  const producer = new Producer(new KafkaClient({ kafkaHost: 'localhost:29092' }));

  await new Promise(resolve => producer.on('ready', resolve));

  return await new Promise((resolve, reject) =>
    producer.createTopics(topics, false, error => {
      if (error) {
        reject(error);
      } else {
        setTimeout(() => producer.close(resolve), 1000);
      }
    }),
  );
};

const stopStreamOnCount = (max: number, stream: Readable) => {
  let current = 0;
  stream.on('data', () => {
    current += 1;
    if (current >= max) {
      stream.push(null);
    }
  });
  return stream;
};

const files = readdirSync(join(__dirname, './assets'));
const sourceData = files.map(file => JSON.parse(String(readFileSync(join(__dirname, './assets', file)))));
const unqiueSourceData = sourceData.map(item => ({ ...item, topic: uuid.v4() }));
const messagesCount = sourceData.reduce((sum, item) => sum + item.messages.length, 0);

describe('Integration test', () => {
  it('Test Serialier', async () => {
    const sourceStream = new ReadableMock(sourceData, { objectMode: true });
    const sinkStream = new WritableMock({ objectMode: true });
    const serializer = new AvroSerializer('http://localhost:8081', {
      logicalTypes: { date: DateType, 'timestamp-millis': TimestampType },
    });

    sourceStream.pipe(serializer).pipe(sinkStream);

    await new Promise(resolve => {
      sinkStream.on('finish', async () => {
        for (const [itemIndex, item] of sinkStream.data.entries()) {
          const type = Type.forSchema(item.schema);
          for (const [messageIndex, message] of item.messages.entries()) {
            const messageParts = deconstructMessage(message);
            const schemaMessage = await idToSchema('http://localhost:8081', messageParts.schemaId);
            const content = type.fromBuffer(messageParts.buffer);

            expect(schemaMessage).toEqual(item.schema);
            expect(content).toEqual(sourceData[itemIndex].messages[messageIndex]);
          }
        }

        resolve();
      });
    });
  });

  it('Test Deserializer with kafka', async () => {
    const sourceStream = new ReadableMock(unqiueSourceData, { objectMode: true });
    const sinkStream = new WritableMock({ objectMode: true });
    const topics = unqiueSourceData.map(item => item.topic);

    const deserializer = new AvroDeserializer('http://localhost:8081', {
      logicalTypes: { date: DateType, 'timestamp-millis': TimestampType },
    });
    const serializer = new AvroSerializer('http://localhost:8081', {
      logicalTypes: { date: DateType, 'timestamp-millis': TimestampType },
    });

    await createTopics(topics);

    const consumerStream = new ConsumerGroupStream(
      {
        kafkaHost: 'localhost:29092',
        groupId: `integration`,
        encoding: 'buffer',
        fromOffset: 'earliest',
      },
      topics,
    );
    const producerStream = new ProducerStream({ kafkaClient: { kafkaHost: 'localhost:29092' } });

    stopStreamOnCount(messagesCount, consumerStream);

    consumerStream.pipe(deserializer).pipe(sinkStream);
    sourceStream.pipe(serializer).pipe(producerStream);

    await new Promise(resolve => {
      sinkStream.on('finish', () => {
        for (let index = 0; index < messagesCount; index++) {
          expect(sinkStream.data[index]).toMatchSnapshot({
            topic: expect.any(String),
            timestamp: expect.any(Date),
          });
        }

        producerStream.close();
        consumerStream.close(resolve);
      });
    });
  }, 15000);
});
