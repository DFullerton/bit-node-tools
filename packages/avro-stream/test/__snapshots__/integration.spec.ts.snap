// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test Test Deserializer with kafka 1`] = `
Object {
  "highWaterOffset": 4,
  "key": null,
  "offset": 0,
  "partition": 0,
  "schema": Object {
    "doc": "Signals that a siemens balance and transaction history was migrated to the new balance platform. Triggered by Balance Service",
    "fields": Array [
      Object {
        "name": "metadata",
        "type": Object {
          "doc": "Metadata, to be used in each event class",
          "fields": Array [
            Object {
              "doc": "A globally unique ID for this Kafka message",
              "name": "eventId",
              "type": "string",
            },
            Object {
              "doc": "An ID that can be used to link all the requests and Kafka messages in a given transaction. If you already have a trace token from a previous event/request, you should copy it here. If this is the very start of a transaction, you should generate a fresh trace token and put it here. A UUID is suitable",
              "name": "traceToken",
              "type": "string",
            },
            Object {
              "doc": "A timestamp for when the event was created (in epoch millis)",
              "name": "createdAt",
              "type": Object {
                "logicalType": "timestamp-millis",
                "type": "long",
              },
            },
          ],
          "name": "EventMetadata",
          "namespace": "com.ovoenergy.kafka.common.event",
          "type": "record",
        },
      },
      Object {
        "doc": "7 digit unique identifier for the customer",
        "name": "accountId",
        "type": "string",
      },
    ],
    "name": "BalanceMigrationCompletedEvent",
    "namespace": "uk.co.boostpower.support.kafka.messages",
    "type": "record",
  },
  "timestamp": Any<Date>,
  "topic": Any<String>,
  "value": BalanceMigrationCompletedEvent {
    "accountId": "***REMOVED***",
    "metadata": EventMetadata {
      "createdAt": "2018-05-05T00:00:00.000Z",
      "eventId": "75f49508-1328-4083-8613-queued",
      "traceToken": "8115ef8a-71c8-4814-aea4-0b7984eae263",
    },
  },
}
`;

exports[`Integration test Test Deserializer with kafka 2`] = `
Object {
  "highWaterOffset": 4,
  "key": null,
  "offset": 1,
  "partition": 0,
  "schema": Object {
    "doc": "Signals that a siemens balance and transaction history was migrated to the new balance platform. Triggered by Balance Service",
    "fields": Array [
      Object {
        "name": "metadata",
        "type": Object {
          "doc": "Metadata, to be used in each event class",
          "fields": Array [
            Object {
              "doc": "A globally unique ID for this Kafka message",
              "name": "eventId",
              "type": "string",
            },
            Object {
              "doc": "An ID that can be used to link all the requests and Kafka messages in a given transaction. If you already have a trace token from a previous event/request, you should copy it here. If this is the very start of a transaction, you should generate a fresh trace token and put it here. A UUID is suitable",
              "name": "traceToken",
              "type": "string",
            },
            Object {
              "doc": "A timestamp for when the event was created (in epoch millis)",
              "name": "createdAt",
              "type": Object {
                "logicalType": "timestamp-millis",
                "type": "long",
              },
            },
          ],
          "name": "EventMetadata",
          "namespace": "com.ovoenergy.kafka.common.event",
          "type": "record",
        },
      },
      Object {
        "doc": "7 digit unique identifier for the customer",
        "name": "accountId",
        "type": "string",
      },
    ],
    "name": "BalanceMigrationCompletedEvent",
    "namespace": "uk.co.boostpower.support.kafka.messages",
    "type": "record",
  },
  "timestamp": Any<Date>,
  "topic": Any<String>,
  "value": BalanceMigrationCompletedEvent {
    "accountId": "***REMOVED***",
    "metadata": EventMetadata {
      "createdAt": "2018-05-05T00:00:00.000Z",
      "eventId": "75f49508-1328-4083-8613-queued",
      "traceToken": "8115ef8a-71c8-4814-aea4-0b7984eae263",
    },
  },
}
`;

exports[`Integration test Test Deserializer with kafka 3`] = `
Object {
  "highWaterOffset": 4,
  "key": null,
  "offset": 2,
  "partition": 0,
  "schema": Object {
    "doc": "Signals that a siemens balance and transaction history was migrated to the new balance platform. Triggered by Balance Service",
    "fields": Array [
      Object {
        "name": "metadata",
        "type": Object {
          "doc": "Metadata, to be used in each event class",
          "fields": Array [
            Object {
              "doc": "A globally unique ID for this Kafka message",
              "name": "eventId",
              "type": "string",
            },
            Object {
              "doc": "An ID that can be used to link all the requests and Kafka messages in a given transaction. If you already have a trace token from a previous event/request, you should copy it here. If this is the very start of a transaction, you should generate a fresh trace token and put it here. A UUID is suitable",
              "name": "traceToken",
              "type": "string",
            },
            Object {
              "doc": "A timestamp for when the event was created (in epoch millis)",
              "name": "createdAt",
              "type": Object {
                "logicalType": "timestamp-millis",
                "type": "long",
              },
            },
          ],
          "name": "EventMetadata",
          "namespace": "com.ovoenergy.kafka.common.event",
          "type": "record",
        },
      },
      Object {
        "doc": "7 digit unique identifier for the customer",
        "name": "accountId",
        "type": "string",
      },
    ],
    "name": "BalanceMigrationCompletedEvent",
    "namespace": "uk.co.boostpower.support.kafka.messages",
    "type": "record",
  },
  "timestamp": Any<Date>,
  "topic": Any<String>,
  "value": BalanceMigrationCompletedEvent {
    "accountId": "***REMOVED***",
    "metadata": EventMetadata {
      "createdAt": "2018-05-05T00:00:00.000Z",
      "eventId": "75f49508-1328-4083-8613-queued",
      "traceToken": "8115ef8a-71c8-4814-aea4-0b7984eae263",
    },
  },
}
`;

exports[`Integration test Test Deserializer with kafka 4`] = `
Object {
  "highWaterOffset": 4,
  "key": null,
  "offset": 3,
  "partition": 0,
  "schema": Object {
    "doc": "Signals that a siemens balance and transaction history was migrated to the new balance platform. Triggered by Balance Service",
    "fields": Array [
      Object {
        "name": "metadata",
        "type": Object {
          "doc": "Metadata, to be used in each event class",
          "fields": Array [
            Object {
              "doc": "A globally unique ID for this Kafka message",
              "name": "eventId",
              "type": "string",
            },
            Object {
              "doc": "An ID that can be used to link all the requests and Kafka messages in a given transaction. If you already have a trace token from a previous event/request, you should copy it here. If this is the very start of a transaction, you should generate a fresh trace token and put it here. A UUID is suitable",
              "name": "traceToken",
              "type": "string",
            },
            Object {
              "doc": "A timestamp for when the event was created (in epoch millis)",
              "name": "createdAt",
              "type": Object {
                "logicalType": "timestamp-millis",
                "type": "long",
              },
            },
          ],
          "name": "EventMetadata",
          "namespace": "com.ovoenergy.kafka.common.event",
          "type": "record",
        },
      },
      Object {
        "doc": "7 digit unique identifier for the customer",
        "name": "accountId",
        "type": "string",
      },
    ],
    "name": "BalanceMigrationCompletedEvent",
    "namespace": "uk.co.boostpower.support.kafka.messages",
    "type": "record",
  },
  "timestamp": Any<Date>,
  "topic": Any<String>,
  "value": BalanceMigrationCompletedEvent {
    "accountId": "***REMOVED***",
    "metadata": EventMetadata {
      "createdAt": "2018-05-05T00:00:00.000Z",
      "eventId": "75f49508-1328-4083-8613-queued",
      "traceToken": "8115ef8a-71c8-4814-aea4-0b7984eae263",
    },
  },
}
`;

exports[`Integration test Test Deserializer with kafka 5`] = `
Object {
  "highWaterOffset": 2,
  "key": null,
  "offset": 0,
  "partition": 0,
  "schema": Object {
    "doc": "Signals that a siemens account migration has been initiated. Triggered by Migration Service",
    "fields": Array [
      Object {
        "name": "metadata",
        "type": Object {
          "doc": "Metadata, to be used in each event class",
          "fields": Array [
            Object {
              "doc": "A globally unique ID for this Kafka message",
              "name": "eventId",
              "type": "string",
            },
            Object {
              "doc": "An ID that can be used to link all the requests and Kafka messages in a given transaction. If you already have a trace token from a previous event/request, you should copy it here. If this is the very start of a transaction, you should generate a fresh trace token and put it here. A UUID is suitable",
              "name": "traceToken",
              "type": "string",
            },
            Object {
              "doc": "A timestamp for when the event was created (in epoch millis)",
              "name": "createdAt",
              "type": Object {
                "logicalType": "timestamp-millis",
                "type": "long",
              },
            },
          ],
          "name": "EventMetadata",
          "namespace": "com.ovoenergy.kafka.common.event",
          "type": "record",
        },
      },
      Object {
        "doc": "7 digit unique identifier for the customer",
        "name": "accountId",
        "type": "string",
      },
      Object {
        "doc": "The unique national reference for a Metering System",
        "name": "mpan",
        "type": "string",
      },
      Object {
        "doc": "The unique identifier used to identify a meter",
        "name": "msn",
        "type": "string",
      },
      Object {
        "doc": "The date when the customer came on supply with Boost (in epoch days)",
        "name": "supplyStartDate",
        "type": Object {
          "logicalType": "date",
          "type": "int",
        },
      },
      Object {
        "doc": "The date when the account enrolled for the new balance platform (in epoch days)",
        "name": "effectiveEnrollmentDate",
        "type": Object {
          "logicalType": "date",
          "type": "int",
        },
      },
    ],
    "name": "AccountMigrationScheduledEvent",
    "namespace": "uk.co.boostpower.support.kafka.messages",
    "type": "record",
  },
  "timestamp": Any<Date>,
  "topic": Any<String>,
  "value": AccountMigrationScheduledEvent {
    "accountId": "***REMOVED***",
    "effectiveEnrollmentDate": "2018-05-05T00:00:00.000Z",
    "metadata": EventMetadata {
      "createdAt": "2018-05-05T00:00:00.000Z",
      "eventId": "75f49508-1328-4083-8613-queued",
      "traceToken": "8115ef8a-71c8-4814-aea4-0b7984eae263",
    },
    "mpan": "6893550d-4221-40ff-a87e-02d8f8fe706d",
    "msn": "10.0",
    "supplyStartDate": "2018-06-05T00:00:00.000Z",
  },
}
`;

exports[`Integration test Test Deserializer with kafka 6`] = `
Object {
  "highWaterOffset": 2,
  "key": null,
  "offset": 1,
  "partition": 0,
  "schema": Object {
    "doc": "Signals that a siemens account migration has been initiated. Triggered by Migration Service",
    "fields": Array [
      Object {
        "name": "metadata",
        "type": Object {
          "doc": "Metadata, to be used in each event class",
          "fields": Array [
            Object {
              "doc": "A globally unique ID for this Kafka message",
              "name": "eventId",
              "type": "string",
            },
            Object {
              "doc": "An ID that can be used to link all the requests and Kafka messages in a given transaction. If you already have a trace token from a previous event/request, you should copy it here. If this is the very start of a transaction, you should generate a fresh trace token and put it here. A UUID is suitable",
              "name": "traceToken",
              "type": "string",
            },
            Object {
              "doc": "A timestamp for when the event was created (in epoch millis)",
              "name": "createdAt",
              "type": Object {
                "logicalType": "timestamp-millis",
                "type": "long",
              },
            },
          ],
          "name": "EventMetadata",
          "namespace": "com.ovoenergy.kafka.common.event",
          "type": "record",
        },
      },
      Object {
        "doc": "7 digit unique identifier for the customer",
        "name": "accountId",
        "type": "string",
      },
      Object {
        "doc": "The unique national reference for a Metering System",
        "name": "mpan",
        "type": "string",
      },
      Object {
        "doc": "The unique identifier used to identify a meter",
        "name": "msn",
        "type": "string",
      },
      Object {
        "doc": "The date when the customer came on supply with Boost (in epoch days)",
        "name": "supplyStartDate",
        "type": Object {
          "logicalType": "date",
          "type": "int",
        },
      },
      Object {
        "doc": "The date when the account enrolled for the new balance platform (in epoch days)",
        "name": "effectiveEnrollmentDate",
        "type": Object {
          "logicalType": "date",
          "type": "int",
        },
      },
    ],
    "name": "AccountMigrationScheduledEvent",
    "namespace": "uk.co.boostpower.support.kafka.messages",
    "type": "record",
  },
  "timestamp": Any<Date>,
  "topic": Any<String>,
  "value": AccountMigrationScheduledEvent {
    "accountId": "***REMOVED***",
    "effectiveEnrollmentDate": "2018-02-05T00:00:00.000Z",
    "metadata": EventMetadata {
      "createdAt": "2018-05-05T00:00:00.000Z",
      "eventId": "95f49508-1328-4083-8613-queued",
      "traceToken": "9115ef8a-71c8-4814-aea4-0b7984eae263",
    },
    "mpan": "223344123123",
    "msn": "12.0",
    "supplyStartDate": "2018-06-05T00:00:00.000Z",
  },
}
`;
